metabulo/views.py

```diff
@@ -86,6 +89,55 @@ def upload_csv_file(file, meta):
         raise
 
 
+_excel_mime_types = [
+    'application/vnd.ms-excel',
+    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
+]
+
+
+@csv_bp.route('/excel/upload', methods=['POST'])
+@use_kwargs({
+    'file': fields.Field(location='files', required=True,
+                         validate=lambda f: f.content_type in _excel_mime_types),
+    'meta': JSONDictStr(missing={})
+})
+def upload_excel_file(file: FileStorage, meta: Dict):
+    excel_sheets = pandas.read_excel(file, sheet_name=None)  # type: Dict[str, pandas.DataFrame]
+    excel_sheets = {sheet: data for (sheet, data) in excel_sheets.items() if not data.empty}
+
+    basename = PurePath(cast(str, file.filename)).with_suffix('')
+
+    try:
+        db_files = []
+
+        def push_file(name: PurePath, data: pandas.DataFrame):
+            db_file = csv_file_schema.load({
+                'name': name.with_suffix('.csv').name,
+                'table': data.to_csv(index=False),
+                'meta': meta
+            })
+
+            db_files.append(db_file)
+            db.session.add(db_file)
+
+        if len(excel_sheets) == 1:
+            # single file case, simple name
+            data = next(iter(excel_sheets.values()))
+            push_file(basename, data)
+        else:
+            for sheet, data in excel_sheets.items():
+                push_file(basename.with_name('%s-%s' % (basename.name, sheet)), data)
+        db.session.commit()
+
+        return jsonify([_serialize_csv_file(f) for f in db_files]), 201
```

Comment:
I usually like to serialize before committing so if there is an error the database gets cleaned up.

Reply:
make sense, I copied the structure from `upload_csv_file`

Reply:
done

Reply:
note: I had to undo it, cause the serialization didn't work properly (like it couldn't identify any columns anymore)

Reply:
Does it work if you run `db.session.flush()` before serialization?